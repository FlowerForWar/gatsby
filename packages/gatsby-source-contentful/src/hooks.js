import { useGatsbyImage, EVERY_BREAKPOINT } from "gatsby-plugin-image"

import { generateImageSource } from "./extend-node-type"

export function useContentfulImage({
  layout = `constrained`,
  baseUrl,
  width,
  height,
  aspectRatio,
  maxWidth,
  ...props
}) {
  const breakpoints = maxWidth
    ? EVERY_BREAKPOINT.filter(bp => bp <= maxWidth)
    : EVERY_BREAKPOINT

  if (layout === `fullWidth`) {
    width ||= maxWidth || breakpoints[breakpoints.length - 1]
    height ||= aspectRatio ? width / aspectRatio : width / (4 / 3)
  } else {
    if (!width || (!height && !aspectRatio)) {
      throw new Error(
        `You must provide width and height for images where layout is ${layout}`
      )
    }
    if (aspectRatio) {
      height ||= width / aspectRatio
    }
  }

  return useGatsbyImage({
    ...props,
    pluginName: `useContentfulImage`,
    generateImageSource,
    layout,
    filename: baseUrl,
    formats: [`auto`],
    breakpoints,
    sourceMetadata: {
      width,
      height,
      format: `auto`,
    },
  })
}
