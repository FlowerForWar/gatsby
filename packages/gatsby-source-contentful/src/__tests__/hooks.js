import { useContentfulImage } from "../hooks"

describe(`useContentfulImage`, () => {
  const consoleWarnSpy = jest.spyOn(console, `warn`)

  const baseUrl = `https://images.ctfassets.net/foo/bar/baz/image.jpg`

  beforeEach(() => {
    consoleWarnSpy.mockClear()
  })
  test(`incomplete config`, () => {
    expect(() => useContentfulImage({ baseUrl })).toThrowError(
      `You must provide width and height for images where layout is constrained`
    )
    expect(consoleWarnSpy).toHaveBeenCalledTimes(0)
  })

  test(`constrained only width`, () => {
    expect(() => useContentfulImage({ baseUrl, width: 600 })).toThrowError(
      `You must provide width and height for images where layout is constrained`
    )
    expect(consoleWarnSpy).toHaveBeenCalledTimes(0)
  })

  test(`constrained: width and aspectRatio`, () => {
    const result = useContentfulImage({
      baseUrl,
      width: 600,
      aspectRatio: 1.778,
    })
    expect(result).toMatchSnapshot()
    expect(consoleWarnSpy).toHaveBeenCalledTimes(0)
  })

  test(`constrained: width and height`, () => {
    const result = useContentfulImage({
      baseUrl,
      width: 600,
      height: 480,
    })
    expect(result).toMatchSnapshot()
    expect(consoleWarnSpy).toHaveBeenCalledTimes(0)
  })

  test(`fixed: width and height`, () => {
    const result = useContentfulImage({
      layout: `fixed`,
      baseUrl,
      width: 600,
      height: 480,
    })
    expect(result).toMatchSnapshot()
    expect(consoleWarnSpy).toHaveBeenCalledTimes(0)
  })

  test(`fullWidth: aspect ratio`, () => {
    const result = useContentfulImage({
      baseUrl,
      layout: `fullWidth`,
      aspectRatio: 1.778,
    })
    expect(result).toMatchSnapshot()
  })

  test(`fullWidth: aspectRatio, maxWidth`, () => {
    const result = useContentfulImage({
      baseUrl,
      layout: `fullWidth`,
      aspectRatio: 1.778,
      maxWidth: 1280,
    })
    expect(result).toMatchSnapshot()
  })
  test(`fullWidth: width, height & maxWidth`, () => {
    const result = useContentfulImage({
      baseUrl,
      layout: `fullWidth`,
      width: 800,
      height: 600,
      maxWidth: 1280,
    })
    expect(result).toMatchSnapshot()
  })
})

// export type Fit = "cover" | "fill" | "inside" | "outside" | "contain"

// export type Layout = "fixed" | "fullWidth" | "constrained"
// export type ImageFormat = "jpg" | "png" | "webp" | "avif" | "auto" | ""
