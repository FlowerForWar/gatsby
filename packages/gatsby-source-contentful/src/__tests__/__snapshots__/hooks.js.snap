// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useContentfulImage constrained: width and aspectRatio 1`] = `
Object {
  "height": 337.4578177727784,
  "images": Object {
    "fallback": Object {
      "sizes": "(min-width: 600px) 600px, 100vw",
      "src": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=600&h=337&fm=auto",
      "srcSet": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=320&h=180&fm=auto 320w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=600&h=337&fm=auto 600w",
    },
    "sources": Array [],
  },
  "layout": "constrained",
  "width": 600,
}
`;

exports[`useContentfulImage constrained: width and height 1`] = `
Object {
  "height": 480,
  "images": Object {
    "fallback": Object {
      "sizes": "(min-width: 600px) 600px, 100vw",
      "src": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=600&h=480&fm=auto",
      "srcSet": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=320&h=256&fm=auto 320w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=600&h=480&fm=auto 600w",
    },
    "sources": Array [],
  },
  "layout": "constrained",
  "width": 600,
}
`;

exports[`useContentfulImage fixed: width and height 1`] = `
Object {
  "height": 320,
  "images": Object {
    "fallback": Object {
      "sizes": "400px",
      "src": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=400&h=320&fm=auto",
      "srcSet": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=400&h=320&fm=auto 400w",
    },
    "sources": Array [],
  },
  "layout": "fixed",
  "width": 400,
}
`;

exports[`useContentfulImage fullWidth: aspect ratio 1`] = `
Object {
  "height": 0.562429696287964,
  "images": Object {
    "fallback": Object {
      "sizes": "100vw",
      "src": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=320&h=180&fm=auto",
      "srcSet": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=320&h=180&fm=auto 320w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=654&h=368&fm=auto 654w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=768&h=432&fm=auto 768w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=1024&h=576&fm=auto 1024w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=1366&h=768&fm=auto 1366w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=1600&h=900&fm=auto 1600w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=1920&h=1080&fm=auto 1920w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=2048&h=1152&fm=auto 2048w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=2560&h=1440&fm=auto 2560w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=3440&h=1935&fm=auto 3440w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=3840&h=2160&fm=auto 3840w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=4096&h=2304&fm=auto 4096w",
    },
    "sources": Array [],
  },
  "layout": "fullWidth",
  "width": 1,
}
`;

exports[`useContentfulImage fullWidth: aspectRatio, maxWidth 1`] = `
Object {
  "height": 0.562429696287964,
  "images": Object {
    "fallback": Object {
      "sizes": "100vw",
      "src": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=320&h=180&fm=auto",
      "srcSet": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=320&h=180&fm=auto 320w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=654&h=368&fm=auto 654w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=768&h=432&fm=auto 768w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=1024&h=576&fm=auto 1024w",
    },
    "sources": Array [],
  },
  "layout": "fullWidth",
  "width": 1,
}
`;

exports[`useContentfulImage fullWidth: width, height & maxWidth 1`] = `
Object {
  "height": 0.75,
  "images": Object {
    "fallback": Object {
      "sizes": "100vw",
      "src": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=800&h=600&fm=auto",
      "srcSet": "https://images.ctfassets.net/foo/bar/baz/image.jpg?w=320&h=240&fm=auto 320w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=654&h=491&fm=auto 654w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=768&h=576&fm=auto 768w,
https://images.ctfassets.net/foo/bar/baz/image.jpg?w=800&h=600&fm=auto 800w",
    },
    "sources": Array [],
  },
  "layout": "fullWidth",
  "width": 1,
}
`;
